optimum/gptq/quantizer.py

```diff
@@ -0,0 +1,571 @@
+            layer = getattr(module, attr)
+            name1 = name + "." + attr if name != "" else attr
+            if name1 in names:
+                device = get_device(layer)
+                delattr(module, attr)
+                if isinstance(layer, nn.Linear):
+                    in_features = layer.in_features
+                    out_features = layer.out_features
+                elif isinstance(layer, nn.Conv2d):
+                    in_features = layer.in_channels
+                    out_features = layer.out_channels
+                elif isinstance(layer, Conv1D):
+                    in_features = layer.weight.shape[0]
+                    out_features = layer.weight.shape[1]
+                if not (self.desc_act) or self.group_size == -1:
+                    new_layer = QuantLinear(
+                        self.bits, self.group_size, in_features, out_features, True, use_cuda_fp16=self.use_cuda_fp16
+                    )
+                else:
+                    new_layer = QuantLinear(self.bits, self.group_size, in_features, out_features, True)
+                new_layer.device = device
+                setattr(module, attr, new_layer.to(device))
+        for name1, child in module.named_children():
+            self._replace_by_quant_layers(child, names, name + "." + name1 if name != "" else name1)
+
+    @torch.no_grad()
+    def quantize_model(
+        self,
+        model: nn.Module,
+        tokenizer: Any,
+        dataset: Union[List[str], str],
+        batch_size: int = 1,
+        pad_token_id: Optional[int] = None,
+    ):
+        """
+        Quantize the model using the dataset
+
+        Args:
+            model (`nn.Module`):
+                The model to quantize
+            dataset (`Union[List[str],str]`):
+                The dataset used for quantization. You can provide your own dataset in a list of string or just use the original datasets used
+                in the paper ['wikitext2','c4'].
+            tokenizer (`Any`, defaults to `None`):
+                The tokenizer to use in order to prepare the dataset
+            batch_size (`Optional[int]`, *optional*, defaults to `1`):
+                The batch size of the dataset
+            pad_token_id (`Optional[int]`, *optional*, defaults to `None`):
+                The pad token id. Needed to prepare the dataset when `batch_size` > 1.
+
+        Returns:
+            `nn.Module`: The quantized model
+        """
+
+        if not torch.cuda.is_available():
+            raise RuntimeError("No GPU found. A GPU is needed to quantize model.")
+
+        model.eval()
+
+        # For Transformer model
+        has_config = False
+        if hasattr(model, "config"):
+            has_config = True
+        if hasattr(model, "hf_device_map"):
+            # If the model has a device_map, we don't move to model. We have already dispatch the hook that will do the work
+            has_device_map = True
+        device = get_device(model)
```

Comment:
This will break in case all the weights of the model are not on the same device as the first weight.

Reply:
If the user don't use accelerate, can we make the assumption that the model is only on one device ?